src/base/net/downloadmanager.cpp

```diff
@@ -105,89 +105,111 @@ namespace
     };
 }
 
-using namespace Net;
+Net::DownloadManager *Net::DownloadManager::m_instance = nullptr;
 
-DownloadManager *DownloadManager::m_instance = nullptr;
-
-DownloadManager::DownloadManager(QObject *parent)
+Net::DownloadManager::DownloadManager(QObject *parent)
     : QObject(parent)
 {
 #ifndef QT_NO_OPENSSL
     connect(&m_networkManager, &QNetworkAccessManager::sslErrors, this, &Net::DownloadManager::ignoreSslErrors);
 #endif
+    connect(&m_networkManager, &QNetworkAccessManager::finished, this, &DownloadManager::handleReplyFinished);
     m_networkManager.setCookieJar(new NetworkCookieJar(this));
 }
 
-void DownloadManager::initInstance()
+void Net::DownloadManager::initInstance()
 {
     if (!m_instance)
         m_instance = new DownloadManager;
 }
 
-void DownloadManager::freeInstance()
+void Net::DownloadManager::freeInstance()
 {
     if (m_instance) {
         delete m_instance;
         m_instance = nullptr;
     }
 }
 
-DownloadManager *DownloadManager::instance()
+Net::DownloadManager *Net::DownloadManager::instance()
 {
     return m_instance;
 }
 
-DownloadHandler *DownloadManager::downloadUrl(const QString &url, bool saveToFile, qint64 limit, bool handleRedirectToMagnet, const QString &userAgent)
+Net::DownloadHandler *Net::DownloadManager::downloadUrl(const QString &url, bool saveToFile, qint64 limit, bool handleRedirectToMagnet, const QString &userAgent)
+{
+    return download({url, saveToFile, limit, handleRedirectToMagnet, userAgent});
+}
+
+Net::DownloadHandler *Net::DownloadManager::download(const DownloadRequest &downloadRequest)
 {
     // Update proxy settings
     applyProxySettings();
 
     // Process download request
-    qDebug("url is %s", qUtf8Printable(url));
-    const QUrl qurl = QUrl(url);
-    QNetworkRequest request(qurl);
+    QNetworkRequest request {downloadRequest.url};
 
-    if (userAgent.isEmpty())
+    if (downloadRequest.userAgent.isEmpty())
         request.setRawHeader("User-Agent", DEFAULT_USER_AGENT);
     else
-        request.setRawHeader("User-Agent", userAgent.toUtf8());
+        request.setRawHeader("User-Agent", downloadRequest.userAgent.toUtf8());
 
     // Spoof HTTP Referer to allow adding torrent link from Torcache/KickAssTorrents
     request.setRawHeader("Referer", request.url().toEncoded().data());
-
-    qDebug("Downloading %s...", request.url().toEncoded().data());
-    qDebug() << "Cookies:" << m_networkManager.cookieJar()->cookiesForUrl(request.url());
-    // accept gzip
+    // Accept gzip
     request.setRawHeader("Accept-Encoding", "gzip");
-    return new DownloadHandler(m_networkManager.get(request), this, saveToFile, limit, handleRedirectToMagnet);
+
+    const QUrl url = request.url();
+    const ServiceID id {url.host(), url.port(80)};
```

Comment:
It is more for those cases when we do not have a QUrl object (i.e. we have string representation of URL and we don't need QUrl instance for anything else). Then `fromUrl()` prevents from creating of QUrl variable.
In cases you meantioned we have QUrl variable (or create it not only for generating ServiceID) so `fromUrl()` doesn't make much sense here.
But I can use it if you have reason to insist.

Reply:
Sorry, I can't understand that. Here is the function signature: `static ServiceID fromURL(const QUrl &url)`. Obviously, a `QUrl` object has to be has to exist to call the function.

Reply:
QUrl has implicit constructor from QString.

Reply:
So what?

Reply:
So we can write:
```c++
auto id = ServiceID::fromURL(urlString);
```
instead of
```c++
const QUrl url {urlString};
ServiceID id {url.host(), url.port(80)};
```
`ServiceID::fromURL()` is just a little helper as I said earlier.

Reply:
Yes, of course. How does it excuse code duplication?

Reply:
Isn't it paranoid to worry about duplication in such a paltry and trivial piece of code? What's the real evil here? I don't even want to discuss it... 
What should I do to calm you down - remove this helper or use it where you have indicated?

Reply:
Replace `{url.host(), url.port(80)}` with ` = ServiceID::fromURL(request.url())`

Reply:
Ok.